File: sorting-searching/sorting-algorithms/bucket-sort/bucket.js
function bucketSort(arr) {
  const n = arr.length;
  const buckets = Array.from({ length: n }, () => []);

  for (const num of arr) {
    const index = Math.min(n - 1, Math.floor(num * n));
    buckets[index].push(num);
  }

  for (const bucket of buckets) {
    bucket.sort((a, b) => a - b);
  }

  let k = 0;
  for (const bucket of buckets) {
    for (const num of bucket) {
      arr[k++] = num;
    }
  }

  return arr;
}

const arr = [0.78, 0.17, 0.39, 0.26, 0.72, 0.94, 0.21, 0.12, 0.23, 0.68];
console.log(bucketSort(arr));

File: sorting-searching/sorting-algorithms/bucket-sort/bucket.py
def bucket_sort(arr):
    n = len(arr)
    buckets = [[] for _ in range(n)]

    for num in arr:
        index = min(int(num * n), n - 1)
        buckets[index].append(num)

    for bucket in buckets:
        bucket.sort()

    result = []
    for bucket in buckets:
        result.extend(bucket)

    return result


if __name__ == "__main__":
    arr = [0.78, 0.17, 0.39, 0.26, 0.72, 0.94, 0.21, 0.12, 0.23, 0.68]
    print("Original:", arr)
    print("Sorted:", bucket_sort(arr))